### PR TITLE
[Cosmos] Computed properties support

### PR DESCRIPTION
**Pull request by balalakshmin**
# Description

Adding computed properties support in `CosmosContainerProperties`.
Corresponding [.Net PR](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3761)

### Usage 

```
        List<ComputedProperty> computedProperties = new ArrayList<>(
            Arrays.asList(
                new ComputedProperty("lowerName", "SELECT VALUE LOWER(IS_DEFINED(c.lastName) ? c.lastName : c.parents[0].familyName) FROM c"),
                new ComputedProperty("upperName", "SELECT VALUE UPPER(IS_DEFINED(c.lastName) ? c.lastName : c.parents[0].familyName) FROM c")
            )
        );

        containerProperties.setComputedProperties(computedProperties);
        CosmosContainerResponse containerResponse = createdDatabase.createContainer(containerProperties);
```

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ContainerPropertiesTest.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
+package com.azure.cosmos;
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ContainerPropertiesTest.java:1]: missing license header

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ContainerPropertiesTest.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.UUID;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Ignore
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ContainerPropertiesTest.java:26]: why are we ignoring it? Is the support not yet there for emulator?

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/TestSuiteBase.java [hunk 3]
```diff
@@ -37,6 +37,7 @@
 import com.azure.cosmos.models.ChangeFeedPolicy;
 import com.azure.cosmos.models.CompositePath;
 import com.azure.cosmos.models.CompositePathSortOrder;
+import com.azure.cosmos.models.ComputedProperty;
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/TestSuiteBase.java:40]: nit: unused imports

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ComputedProperty.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ */
+public final class ComputedProperty {
+    private String name;
+    private String query;
+    private JsonSerializable jsonSerializable;
+
+    /**
+     * Instantiates a new Computed properties.
+     */
+    public ComputedProperty() {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ComputedProperty.java:19]: Does this needs to be public? Can users create a ComputedProperty instance without name and query?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosContainerProperties.java [hunk 5]
```diff
@@ -178,6 +179,27 @@ public CosmosContainerProperties setChangeFeedPolicy(ChangeFeedPolicy value) {
...
+     * Gets the computedProperties for this container in the Azure Cosmos DB service.
+     *
+     * @return the computedProperties.
+     */
+    public Collection<ComputedProperty> getComputedProperties() {
+        return this.documentCollection.getComputedProperties();
+    }
+
+    /**
+     * Sets the computedProperties for this container in the Azure Cosmos DB service.
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosContainerProperties.java:193]: Is there more documentation available for this? like what ComputedProperty is, and what it does, how can customers use it, etc. Or even link to a public doc if possible

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ComputedProperty.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        return name;
+    }
+
+    /**
+     * Gets the query used to evaluate the value for the computed property.
+     *
+     * @return the query for the computed property.
+     */
+    public String getQuery() {
+        return query;
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ComputedProperty.java:66]: How do you ever propagate name and query in the Json serializable? I think there are some gaps in the JsonSerializable implementation here. Make sure to test creating a new instance with name and query, serializing it, deseializing the json serializied before and compare the results etc.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosContainerProperties.java [hunk 7]
```diff
@@ -178,6 +179,27 @@ public CosmosContainerProperties setChangeFeedPolicy(ChangeFeedPolicy value) {
...
+     *
+     * @param computedProperties the computedProperties.
+     * @return the CosmosContainerProperties.
+     */
+    public CosmosContainerProperties setComputedProperties(Collection<ComputedProperty> computedProperties) {
+        this.documentCollection.setComputedProperties(computedProperties);
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosContainerProperties.java:199]: validations missing? Can computedProperties be ever null? What if the list contains null values etc.

**Review by FabianMeiswinkel — CHANGES_REQUESTED** [review]
Mostly blocking in the probably incomplete JsonSerializable implementation.

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ContainerPropertiesTest.java (hunk 2 above)
Comment (balalakshmin) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ContainerPropertiesTest.java:26]: Yes, it's not there in emulator yet.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ComputedProperty.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+// Licensed under the MIT License.
+package com.azure.cosmos.models;
+
+import com.azure.cosmos.implementation.JsonSerializable;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+/**
+ * Represents a computed property definition for a Cosmos DB container.
+ */
+public final class ComputedProperty {
```
[the remark this answers is not in the document]
Comment (balalakshmin) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ComputedProperty.java:11]: Other models similar to this - IndexingPolicy, UniqueKeyPolicy, etc. which are present in CosmosContainerProperties don't have the Cosmos prefix.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ComputedProperty.java (hunk 6 above)
Comment (balalakshmin) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ComputedProperty.java:66]: This is being deserialized as POJO. [Ref](https://github.com/Azure/azure-sdk-for-java/blob/57b125f63d5ea64b2778b823647ba43143a344da/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java#L514). I'll remove the unnecessary methods from this model.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosContainerProperties.java (hunk 7 above)
Comment (balalakshmin) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosContainerProperties.java:199]: The check is inside the documentCollection class. I've added check for the presence of null values also

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ComputedProperty.java (hunk 6 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ComputedProperty.java:46]: Thanks for adding the new test validating json serialization end-to-end

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM now...

File: sdk/cosmos/azure-cosmos/CHANGELOG.md [hunk 9]
```diff
@@ -5,6 +5,8 @@
 #### Features Added
 * Added the capability to filter request-level metrics based on diagnostic thresholds. Request-level metrics usually are used to capture metrics per backend endpoint/replica - a high cardinality dimension. Filtering by diagnostic thresholds reduces the overhead - but also means request-level metrics can only be used for debugging purposes - not for monitoring purposes. So, it is important to use the unfiltered operation-level metrics for health monitoring in this case. - See [PR 35114](https://github.com/Azure/azure-sdk-for-java/pull/35114)
 * Added optional tags/dimensions for PartitionId/ReplicaId as alternative to ServiceAddress for direct-mode (rntbd) request-level metrics. - See [PR 35164](https://github.com/Azure/azure-sdk-for-java/pull/35164) 
+* Added support for computed properties - See [PR 35046](https://github.com/Azure/azure-sdk-for-java/pull/35046)
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/CHANGELOG.md:8]:
```suggestion
* Added support for `ComputedProperty` in `CosmosContainerProperties` - See [PR 35046](https://github.com/Azure/azure-sdk-for-java/pull/35046)
```

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ContainerPropertiesTest.java [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    private CosmosClient client;
+    private CosmosDatabase createdDatabase;
+    private CosmosContainer container;
+
+    @Factory(dataProvider = "clientBuilders")
+    public ContainerPropertiesTest(CosmosClientBuilder clientBuilder) {
+        super(clientBuilder);
+    }
+
+    @BeforeClass(groups = {"emulator"}, timeOut = SETUP_TIMEOUT)
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ContainerPropertiesTest.java:43]: if not supported in emulator, should we use `simple` test groups instead?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ComputedProperty.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+import com.azure.cosmos.implementation.JsonSerializable;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+/**
+ * Represents a computed property definition for a Cosmos DB container.
+ */
+public final class ComputedProperty {
+    @JsonProperty("name")
+    private String name;
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ComputedProperty.java:14]: I believe these variables can be `final`

**Review by kushagraThapar — CHANGES_REQUESTED** [review]
Thanks for the changes @balalakshmin, they look great. 

One new process that we are starting now is to make sure that every feature PR also adds samples / documentation about the feature. 

For this one, I think we should add the code snippet about this feature, in the samples folder, and then reference those code snippets in the java docs on these APIs. 

You can see the example of throughput control / change feed processor, also this sample PR will show you how / where to add the code snippets - https://github.com/Azure/azure-sdk-for-java/pull/35300

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

File: sdk/cosmos/azure-cosmos/src/samples/java/com/azure/cosmos/ComputedPropertiesCodeSnippet.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
+package com.azure.cosmos;
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/samples/java/com/azure/cosmos/ComputedPropertiesCodeSnippet.java:1]: missing new file license header. Can be added later on after CI finishes.

**Review by kushagraThapar — APPROVED** [review]
LGTM, thanks @balalakshmin !

**Pull request merged**